doc/vastd.1
.TH VASTD 1 "September 01, 2015" 0.1 "Visibility Across Space and Time"
.SH NAME
.PP
\fB\fCvastd\fR \-\- VAST daemon running a node actor
.SH SYNOPSIS
.PP
\fB\fCvastd\fR [\fIoptions\fP]
.SH OVERVIEW
.PP
VAST is a distributed platform for large\-scale network forensics. It's modular
system architecture is exclusively implemented in terms of the actor model. In
this model, concurrent entities (actors) execute in parallel and
communicate asynchronously solely via message passing. Users spawn system
components as actors and connect them together to create a custom system
topology.
.SH DESCRIPTION
.PP
The \fB\fCvastd\fR process spawns a special \fInode\fP actor, which is as a container for
other actors. Typically, each physical machine in a VAST deployment runs a
single \fB\fCvastd\fR process. For single\-node deployments all actors run inside this
process, whereas cluster deployments consist of multiple nodes with actors
spread across them.
.PP
Nodes can enter a peering relationship to form a cluster. All peers have the
same authority: if one fails, others can take over. By default, each node
includes all core actors: archive, index, importer, identifier. For more
fine\-grained control about the components running on a node, one can spawn the
node in bare mode to get an empty node. This allows for more flexible
arrangement of components to best match the available system hardware.
.SH OPTIONS
.PP
The following \fIoptions\fP are available:
.TP
\fB\fC\-b\fR
Run in \fIbare\fP mode, i.e., do not spawn any actors. Use \fIbare\fP mode when you
want to create a custom topology. When not specifying this option, \fB\fCvastd\fR
automatically spawns all core actors by executing the following commands
upon spawning the node:
.PP
.RS
.nf
  vast spawn identifier
  vast spawn importer
  vast spawn archive
  vast spawn index
  vast connect importer identifier
  vast connect importer archive
  vast connect importer index
.fi
.RE
.TP
\fB\fC\-d\fR \fIdirectory\fP [\fIvast\fP]
The path on the file system where to store persistent state.
.TP
\fB\fC\-e\fR \fIendpoint\fP [\fI127.0.0.1:42000\fP]
The endpoint of the node to connect to or launch. (See 
.BR vast (1) 
for syntax)
.TP
\fB\fC\-f\fR
Start daemon in foreground, i.e., do not detach from controlling terminal and
run in background.
.TP
\fB\fC\-h\fR
Display a help message and exit.
.TP
\fB\fC\-l\fR \fIverbosity\fP [\fI3\fP]
The logging verbosity. (See 
.BR vast (1) 
for an explanation of values)
.TP
\fB\fC\-m\fR \fImessages\fP [\fI\-1\fP]
The CAF worker throughput expressed in the maximum number of messages to
process when a worker gets scheduled. The default value of \fI\-1\fP means an
unlimited number of messages.
.TP
\fB\fC\-n\fR \fIname\fP [\fIhostname\fP]
Overrides the node \fIname\fP, which defaults to the system hostname. Each node
in an topology must have a unique name, otherwise peering fails.
.TP
\fB\fC\-p\fR \fIlogfile\fP
Enable CAF profiling of worker threads and actors and write the per\-second
sampled data to \fIlogfile\fP\&.
.TP
\fB\fC\-t\fR \fIthreads\fP [\fIstd::thread::hardware_concurrency()\fP]
The number of worker threads to use for CAF's scheduler.
.TP
\fB\fC\-v\fR
Print VAST version and exit.
.SH EXAMPLES
.PP
Start a node in the foreground:
.PP
.RS
.nf
vastd \-f
.fi
.RE
.PP
Start a node with debugging logging verbosity and use directory "test":
.PP
.RS
.nf
vastd \-l 5 \-d test 
.fi
.RE
.PP
Start a node at a different port with 10 worker threads:
.PP
.RS
.nf
vastd \-e :6666 \-t 10
.fi
.RE
.SH BUGS
.PP
If you encounter a bug or have suggestions for improvement, please file an
issue at 
\[la]https://github.com/mavam/vast/issues\[ra]\&.
.SH SEE ALSO
.PP
.BR vast (1)
.PP
Visit 
\[la]https://github.com/mavam/vast\[ra] for more information about VAST.


src/vastd.cc
#include "vast/config.h"

#ifdef VAST_LINUX
#include <unistd.h> // daemon(3)
#endif

#include <csignal>
#include <cstdlib>
#include <cstring> // strsignal(3)

#include <algorithm>
#include <iostream>
#include <limits>
#include <string>
#include <thread>

#include <caf/io/all.hpp>

#include "vast/announce.h"
#include "vast/banner.h"
#include "vast/caf.h"
#include "vast/filesystem.h"
#include "vast/logger.h"
#include "vast/optional.h"
#include "vast/actor/node.h"
#include "vast/actor/signal_monitor.h"
#include "vast/concept/printable/vast/error.h"
#include "vast/detail/adjust_resource_consumption.h"
#include "vast/util/endpoint.h"
#include "vast/util/string.h"
#include "vast/util/system.h"

using namespace vast;
using namespace std::string_literals;

int main(int argc, char* argv[]) {
  if (!detail::adjust_resource_consumption())
    return 1;
  // Defaults.
  auto dir = "vast"s;
  auto endpoint = std::string{};
  auto host = "127.0.0.1"s;
  auto log_level = 3;
  auto messages = std::numeric_limits<size_t>::max();
  auto name = util::split_to_str(util::hostname(), ".")[0];
  auto port = uint16_t{42000};
  auto profile_file = std::string{};
  auto threads = std::thread::hardware_concurrency();
  // Parse and validate command line.
  auto r = message_builder(argv + 1, argv + argc).extract_opts({
    {"bare,b", "spawn empty node without any actors"},
    {"directory,d", "path to persistent state directory", dir},
    {"endpoint,e", "the node endpoint", endpoint},
    {"foreground,f", "run daemon in foreground"},
    {"log-level,l", "verbosity of console and/or log file", log_level},
    {"messages,m", "maximum messages per CAF scheduler invocation", messages},
    {"name,n", "the name of this node", name},
    {"profile,p", "enable CAF profiler", profile_file},
    {"threads,t", "number of worker threads in CAF scheduler", threads},
    {"version,v", "print version and exit"}
  });
  if (! r.error.empty()) {
    std::cerr << r.error << std::endl;
    return 1;
  }
  if (r.opts.count("version") > 0) {
    std::cout << VAST_VERSION << std::endl;
    return 0;
  }
  if (r.opts.count("help") > 0) {
    std::cout << banner() << "\n\n" << r.helptext;
    return 0;
  }
  if (r.opts.count("endpoint") > 0
      && !util::parse_endpoint(endpoint, host, port)) {
    std::cerr << "invalid endpoint: " << endpoint << std::endl;
    return 1;
  }
  if (!r.remainder.empty()) {
    auto arg = r.remainder.get_as<std::string>(0);
    std::cerr << "invalid stray argument: " << arg << std::endl;
    return 1;
  }
  // Initialize logger.
  auto verbosity = static_cast<logger::level>(log_level);
  auto log_dir = path{dir} / node::log_path();
  if (!logger::file(verbosity, (log_dir / "vast.log").str())) {
    std::cerr << "failed to initialize logger file backend" << std::endl;
    return 1;
  }
  if (r.opts.count("foreground")) {
    auto colorized = true;
    if (!logger::console(verbosity, colorized)) {
      std::cerr << "failed to initialize logger console backend" << std::endl;
      return 1;
    }
  } else {
    VAST_DEBUG("deamonizing process (PID", util::process_id() << ")");
    // On Mac OS, daemon(3) is deprecated since 10.5.
    VAST_DIAGNOSTIC_PUSH
    VAST_DIAGNOSTIC_IGNORE_DEPRECATED
    if (::daemon(0, 0) != 0) {
      VAST_ERROR("failed to daemonize process");
      return 1;
    }
    VAST_DIAGNOSTIC_POP
  }
  // Replace/adjust scheduler.
  if (r.opts.count("profile"))
    set_scheduler(
      new scheduler::profiled_coordinator<>{
        profile_file, std::chrono::milliseconds{1000}, threads, messages});
  else if (r.opts.count("threads") || r.opts.count("messages"))
    set_scheduler<>(threads, messages);
  VAST_VERBOSE(banner() << "\n\n");
  VAST_VERBOSE("set scheduler threads to", threads);
  VAST_VERBOSE("set scheduler maximum throughput to",
               (messages == std::numeric_limits<size_t>::max()
                ? "unlimited" : std::to_string(messages)));
  // Enable direct connections.
  VAST_VERBOSE("enabling direct connection optimization");
  auto cfg = whereis(atom("ConfigServ"));
  anon_send(cfg, put_atom::value, "global.enable-automatic-connections",
            make_message(true));
  // Initialize node actor.
  auto guard = make_scope_guard(
    [] { caf::shutdown(); logger::destruct(); }
  );
  announce_types();
  auto node = spawn<vast::node>(name, dir);
  scoped_actor self;
  if (r.opts.count("bare") == 0)
    self->sync_send(node, "spawn", "core").await([](ok_atom) {});
  try {
    // Publish the node.
    auto bound_port = caf::io::publish(node, port, host.c_str());
    VAST_VERBOSE("listening on", host << ':' << bound_port,
                 "with name \"" << name << '"');
    /// Install signal handlers and block until either a signal arrives or the
    /// node terminates.
    auto sig_mon = self->spawn(signal_monitor::make, self);
    self->monitor(node);
    auto stop = false;
    self->do_receive(
      [&](down_msg const& msg) {
        VAST_DEBUG("received DOWN from " << msg.source);
        stop = true;
      },
      [&](signal_atom, int signal) {
        VAST_DEBUG("got " << ::strsignal(signal));
        if (signal == SIGINT || signal == SIGTERM)
          stop = true;
        else
          self->send(node, signal_atom::value, signal);
      },
      others() >> [&] {
        VAST_WARN("received unexpected message:",
                   to_string(self->current_message()));
      }
    ).until([&] { return stop == true; });
    if (node->exit_reason() == exit_reason::not_exited)
      self->send_exit(node, exit::stop);
    self->send_exit(sig_mon, exit::stop);
    self->await_all_other_actors_done();
    auto er = node->exit_reason();
    if (er == exit::error)
      return 1;
    else if (er == exit::kill)
      return -1;
    else if (!(er == exit::done || er == exit::stop))
      return 2;
  } catch (caf::network_error const& e) {
    VAST_ERROR(e.what());
    self->send_exit(node, exit::stop);
    return 1;
  } catch (...) {
    VAST_ERROR("terminating due to uncaught exception");
    return 1;
  }
  return 0;
}
